### PR TITLE
Dynamic Includes

### PR DESCRIPTION
**Pull request by gwincr11**
### Summary

When working on a significantly large application, developers will sink many hours into setting up specific pages to preload data in the correct way across different types of database loading patterns. The complexity involved in many of these methods will have cross cutting effects to pages that do not receive the same level of effort towards optimization. When those other pages access the same data, often by developers working in other areas of the code, they may not have the requisite knowledge to properly optimize the models associations or methods causing Rails to introduce n+1 queries. 

This PR offers a few different patterns to help with these situations. The key feature being dynamic_includes, this is a proposed Rails feature which ties into the existing Preloader, lazily running it on accessed associations that are not preloaded, when the feature is enabled, preventing n+1 queries. Here is an example of what these changes enable from a developer perspective.

### Example Usage

These examples are taken from the tests added to the PR. Without dynamic includes enabled, the following implementation would execute 9 queries:

```ruby
developers = Developer.where(id: [developer.id, developer2.id])
assert_queries(9) do
  developers.each do |d|
    d.ship.parts.each do |part|
      part.trinkets.each do |t|
        t
      end
    end
  end
end
```

<details>
<summary>SQL Log</summary>

<pre>
D, [2022-05-05T16:18:46.432004 #3280] DEBUG -- :   Developer Load (0.2ms)  SELECT "developers"."id", "developers"."name", "developers"."salary", "developers"."firm_id", "developers"."mentor_id", "developers"."legacy_created_at", "developers"."legacy_updated_at", "developers"."legacy_created_on", "developers"."legacy_updated_on" FROM "developers" WHERE "developers"."id" IN (?, ?)  [["id", 1], ["id", 11]]
D, [2022-05-05T16:18:46.432893 #3280] DEBUG -- :   Ship Load (0.1ms)  SELECT "ships".* FROM "ships" WHERE "ships"."developer_id" = ? LIMIT ?  [["developer_id", 1], ["LIMIT", 1]]
D, [2022-05-05T16:18:46.433554 #3280] DEBUG -- :   ShipPart Load (0.1ms)  SELECT "ship_parts".* FROM "ship_parts" WHERE "ship_parts"."ship_id" = ?  [["ship_id", 2]]
D, [2022-05-05T16:18:46.434335 #3280] DEBUG -- :   Treasure Load (0.1ms)  SELECT "treasures".* FROM "treasures" WHERE "treasures"."looter_id" = ? AND "treasures"."looter_type" = ?  [["looter_id", 1], ["looter_type", "ShipPart"]]
D, [2022-05-05T16:18:46.434911 #3280] DEBUG -- :   Treasure Load (0.1ms)  SELECT "treasures".* FROM "treasures" WHERE "treasures"."looter_id" = ? AND "treasures"."looter_type" = ?  [["looter_id", 2], ["looter_type", "ShipPart"]]
D, [2022-05-05T16:18:46.435559 #3280] DEBUG -- :   Ship Load (0.0ms)  SELECT "ships".* FROM "ships" WHERE "ships"."developer_id" = ? LIMIT ?  [["developer_id", 11], ["LIMIT", 1]]
D, [2022-05-05T16:18:46.436479 #3280] DEBUG -- :   ShipPart Load (0.1ms)  SELECT "ship_parts".* FROM "ship_parts" WHERE "ship_parts"."ship_id" = ?  [["ship_id", 751016585]]
D, [2022-05-05T16:18:46.437258 #3280] DEBUG -- :   Treasure Load (0.1ms)  SELECT "treasures".* FROM "treasures" WHERE "treasures"."looter_id" = ? AND "treasures"."looter_type" = ?  [["looter_id", 3], ["looter_type", "ShipPart"]]
D, [2022-05-05T16:18:46.438098 #3280] DEBUG -- :   Treasure Load (0.1ms)  SELECT "treasures".* FROM "treasures" WHERE "treasures"."looter_id" = ? AND "treasures"."looter_type" = ?  [["looter_id", 4], ["looter_type", "ShipPart"]]
</pre>
</details>

When enabling dynamic includes it happens in 4:

```ruby
developers = Developer.where(id: [developer.id, developer2.id])

ActiveRecord.enable_dynamic_includes do
  assert_queries(4) do
    developers.each do |d|
      d.ship.parts.each do |part|
        part.trinkets.each do |t|
          t
        end
      end
    end
  end
end
```

<details>
<summary>SQL Log</summary>

<pre>
D, [2022-05-05T16:18:46.388805 #3280] DEBUG -- :   Developer Load (0.2ms)  SELECT "developers"."id", "developers"."name", "developers"."salary", "developers"."firm_id", "developers"."mentor_id", "developers"."legacy_created_at", "developers"."legacy_updated_at", "developers"."legacy_created_on", "developers"."legacy_updated_on" FROM "developers" WHERE "developers"."id" IN (?, ?)  [["id", 1], ["id", 11]]
D, [2022-05-05T16:18:46.424796 #3280] DEBUG -- :   Ship Load (0.3ms)  SELECT "ships".* FROM "ships" WHERE "ships"."developer_id" IN (?, ?)  [["developer_id", 1], ["developer_id", 11]]
D, [2022-05-05T16:18:46.425470 #3280] DEBUG -- : Dynamically preloaded: Developer.ship
D, [2022-05-05T16:18:46.427809 #3280] DEBUG -- :   ShipPart Load (0.2ms)  SELECT "ship_parts".* FROM "ship_parts" WHERE "ship_parts"."ship_id" IN (?, ?)  [["ship_id", 2], ["ship_id", 751016585]]
D, [2022-05-05T16:18:46.428172 #3280] DEBUG -- : Dynamically preloaded: Developer.ship.parts
D, [2022-05-05T16:18:46.430243 #3280] DEBUG -- :   Treasure Load (0.2ms)  SELECT "treasures".* FROM "treasures" WHERE "treasures"."looter_type" = ? AND "treasures"."looter_id" IN (?, ?, ?, ?)  [["looter_type", "ShipPart"], ["looter_id", 1], ["looter_id", 2], ["looter_id", 3], ["looter_id", 4]]
D, [2022-05-05T16:18:46.430748 #3280] DEBUG -- : Dynamically preloaded: Developer.ship.parts.trinkets
</pre>

</details>

In some scenarios it is ideal to not preload all the data for a particular algorithm, if perhaps it has a very expensive database query associated and perhaps a select runs prior to loading the data to limit the selected data further in memory, you may want to turn off the dynamic includes in such a scenario. This is an exceptional case but is possible. 

The below example will take 7 queries:

```ruby
ActiveRecord.enable_dynamic_includes do
  developers = Developer.where(id: [developer.id, developer2.id])
  assert_queries(7) do
  developers.each do |d|
    assert ActiveRecord.dynamic_includes_enabled?
    ActiveRecord.disable_dynamic_includes do
      d.ship.parts.each do |part| # This should n+1
        assert_not ActiveRecord.dynamic_includes_enabled?
          ActiveRecord.enable_dynamic_includes do
            part.trinkets.each do |t|
              t
              assert ActiveRecord.dynamic_includes_enabled?
            end
         end
         assert_not ActiveRecord.dynamic_includes_enabled?
       end
     end
     assert ActiveRecord.dynamic_includes_enabled?
   end
 end
end
 ```

<details>
<summary>SQL Log</summary>

```
D, [2022-05-05T16:18:46.439329 #3280] DEBUG -- :   Developer Load (0.2ms)  SELECT "developers"."id", "developers"."name", "developers"."salary", "developers"."firm_id", "developers"."mentor_id", "developers"."legacy_created_at", "developers"."legacy_updated_at", "developers"."legacy_created_on", "developers"."legacy_updated_on" FROM "developers" WHERE "developers"."id" IN (?, ?)  [["id", 1], ["id", 11]]
D, [2022-05-05T16:18:46.440002 #3280] DEBUG -- :   Ship Load (0.1ms)  SELECT "ships".* FROM "ships" WHERE "ships"."developer_id" = ? LIMIT ?  [["developer_id", 1], ["LIMIT", 1]]
D, [2022-05-05T16:18:46.440643 #3280] DEBUG -- :   ShipPart Load (0.1ms)  SELECT "ship_parts".* FROM "ship_parts" WHERE "ship_parts"."ship_id" = ?  [["ship_id", 2]]
D, [2022-05-05T16:18:46.441803 #3280] DEBUG -- :   Treasure Load (0.1ms)  SELECT "treasures".* FROM "treasures" WHERE "treasures"."looter_type" = ? AND "treasures"."looter_id" IN (?, ?)  [["looter_type", "ShipPart"], ["looter_id", 1], ["looter_id", 2]]
D, [2022-05-05T16:18:46.442134 #3280] DEBUG -- : Dynamically preloaded: Developer.ship.parts.trinkets
D, [2022-05-05T16:18:46.442986 #3280] DEBUG -- :   Ship Load (0.1ms)  SELECT "ships".* FROM "ships" WHERE "ships"."developer_id" = ? LIMIT ?  [["developer_id", 11], ["LIMIT", 1]]
D, [2022-05-05T16:18:46.443639 #3280] DEBUG -- :   ShipPart Load (0.1ms)  SELECT "ship_parts".* FROM "ship_parts" WHERE "ship_parts"."ship_id" = ?  [["ship_id", 751016585]]
D, [2022-05-05T16:18:46.445137 #3280] DEBUG -- :   Treasure Load (0.2ms)  SELECT "treasures".* FROM "treasures" WHERE "treasures"."looter_type" = ? AND "treasures"."looter_id" IN (?, ?)  [["looter_type", "ShipPart"], ["looter_id", 3], ["looter_id", 4]]
D, [2022-05-05T16:18:46.445454 #3280] DEBUG -- : Dynamically preloaded: Developer.ship.parts.trinkets
```

</details>

### Changes that enable this functionality

This is accomplished through 3 new concepts:

1. Load Tree: This class maintains a tree-like structure of the loaded Active Record objects. 
2. Dynamic Includes: A new ActiveRecord block that when enabled dynamically preloads Rails association on access (lazy) preventing n+1 queries.
3. Overfetch debug mode: A new debugging mode that stores all the loaded and used ActiveRecord objects, and expands the Load Tree to also maintain a call stack for each instantiated object so that developers can easily find where and when an object was instantiated.

These new concepts enable the following new capabilities which are key to the functionality this pr introduces:

1. Load Trees maintains a list of siblings, which are records loaded at the same time, which opens the door to a dynamic includes feature that lazy preloads associations to prevent n+1 queries.
2.  The Load Tree API is public so that other Gems such as [prelude](https://github.com/seejohnrun/prelude) can interface directly with it, expanding the use of Load Tree and allowing for tools like prelude to run a lazy dynamic preload as well. This greatly simplifies the development experience when moving between different data loading patterns, since the api for preloading is extensible. ([Here is a spike that investigated how to do this with prelude](https://github.com/seejohnrun/prelude/compare/main...gwincr11:prelude:cg-load_tree))
3. Dynamic Include blocks can be nested ensuring all method callers load database entities in the most performant way by scoping the appropriate n+1/include behavior to method scopes.
4. Storing the Load Tree allows for a walkable tree of all loaded records, providing useful debugging information and exposing valuable information to other consumers about how records are related.
5. Load Tree provides deep introspection into how each record instance was instantiated allowing developers to use graphical interfaces to reason about what objects were instantiated, how and when.
 

### Findings:

As part of my testing process I set  up the GitHub test suite with dynamic includes enabled globally. This helped me find some edge cases and also let me take a look at what sort of patterns dynamic includes frequently fixed. 

Running the GitHub test suite to completion locally is impossible, as it will eventually time out, so I turned it on for as long as it would run to see how many times the dynamic includes filled an association that was accessed but not preloaded. I was able to get through about 1800 tests and the prefill ran 671 times. I then ran just the integration tests for about 30 minutes, it timed out pretty quickly, showing that it prefilled 248 missed prefills. This was only prefilling associations, I suspect extending it to prelude and twirp calls would further increase this number as we frequently jump between patterns.

There were also a large number of tests that had assertions against how many queries were allowed. After turning on dynamic includes most of these tests were failing, the number of queries had dropped, sometimes by up to 50%. As a part of my testing, I reviewed places in the GitHub code base where query counts dropped significantly, in order to find patterns that the dynamic includes were frequently fixed. The most common fixes I found included:

Destroy_all calls with associations frequently nested multiple layers deep that utilized destroy_dependent association properties.
Update, save, creation or other active record callbacks that manipulate associations after changes occur.
Model access patterns where a method call on the parent instance had subsequent method calls more than 2 layers deep which  loaded an association.
Active Record calls on a model (x) which instantiaed other model instances (y) with ActiveRecord where or find, which was a method and not association, that then loaded an association (z) further down the method chain of the instantiated (y) object.

### Tradeoff to Dynamic Includes:

As with any preload approach there is a risk that data that is not used will be preloaded. In practice after testing on GitHub in a few places, the dynamic includes seem to provide more than enough value to offset any concern about over fetching given the quality concern where less database calls are valued over potentially more memory usage. 

The pattern in which over fetching occurs is worth noting, it is not uncommon to get an array of results back from the database and then select elements that meet some criteria out of the array in memory. In this case, the sibling tree (which is relied upon to preload associations) will not be altered, so selecting a subset of an array and then iterating through it while loading associations will load the associations on records (siblings) that are not in the current iteration (the resulting array of the selection process). 

```ruby
ActiveRecord.enable_dynamic_includes do
  devs = Developers.where(id: [some ids]) 
  staff = devs.select(&:employee?)
  staff.each do |s|
    s.association # this will load the association on all devs
  end
end
```

This can be handled in three ways (that occur to me), easiest is to disable dynamic includes and N+1, another option is to disable dynamic includes around this operation and manually preload for just the selected records.

```ruby
ActiveRecord.enable_dynamic_includes do
  devs = Developers.where(id: [some ids]) 
  staff = devs.select(&:employee?)
  ActiveRecord.disable_dynamic_includes do
    ActiveRecord::Associations::Preloader.new(records: staff, associations: [:association]).call
    staff.each do |s|
      s.association # this will load the association on all devs
    end
  end
end
```

The final option I can think of would be to manually reset the siblings on the selected records load trees. 

```ruby
ActiveRecord.enable_dynamic_includes do
  devs = Developers.where(id: [some ids]) 
  staff = devs.select(&:employee?)
  staff.each do |s|
    s._load_tree.reset_siblings(staff)
  end
  staff.each do |s|
    s.association # this will load the association on all devs
  end
end
```

This pattern, in the GitHub codebase, without dynamic includes, appears to frequently introduce an N+1 or to overfetch.  The `includes()` api that ActiveRecord expose I believe suffers this same problem since it needs to be set up prior to the select occuring, so over fetching is not uncommon with includes today. GitHub has a custom Preload tool that we use in these scenarios, but developers need to recognize this problem to avoid it. Even when overfetching happens the concern that dynamic includes focuses on, less database calls preferred to more memory usage, holds as the overfetching still incurs only 1 database call for the over fetched association.

### Alternative Syntaxes Explored

I tried out 3 different syntaxes for this and landed on the block based function calls for a few reasons. The first version I tried was the classic Rails includes, where the user lists out the includes they want to allow. This syntax was undesirable for a few reasons. It is far too easy to miss an include and introduce an N+1, especially as model interactions become more complex. It also prevents a user from toggling on and off includes, when associations are nested. For example if I have a Blog post with comments and authors and authors have posts. Suppose I want to allow the system to dynamically include comments and posts for authors but I need to do something on authors to determine when to display posts. The closest I can come is this: 
`Blogs.include([comments: [authors: [posts]]])`, which is not what I wish to acheive.

Next I tried chaining the include as part of the Model so `blogs = Blog.all.enable_dynamic_includes.posts`. This would pass the setting to all the children it created for the association `blogs.comments` comments would dynamically prefill. This allowed for the ability to enable and disable the includes in the way I wanted but keeping track of all the state was messy. It was also pretty easy to turn it off then get deep down into model calls and not realize it had been left off, as the state was not intuitive to follow from parent to child.

The block based model proved to be very flexible, easy to understand and allows for the feature to be enabled for entire sections of the codebase without lots of change to the underlying model code. It also has a really nice feature of allowing methods to define their preferred behavior and going back to the calling methods preferred setting once the scope of the called method ends. The other benefit is that it can easily be turned on for a single controller action with a before filter to slowly roll it out and try out the behavior.

```ruby
def dynamic_includes
  ActiveRecord.enable_dynamic_includes do
    yield
  end
end
```

### Debugging Tools

I have put together a template which simplifies debugging how ActiveRecord loads data, it is available on this [gist](https://gist.github.com/gwincr11/35adf30f1892beae2db86f025c340c24). Simply drop it into your Rails template and enable `ActiveRecord.enable_dynamic_includes_overfetch_debugging`, this template will show the full load tree and navigate through parent/association relationships by opening the html summaries. The template works recursively so it is possible to walk through the entire load tree for any record.

Here is a demo video of the debugging tool I created with this to help with the development of this feature, it would be very easy to develop gems or even Rails tooling to enable sophisticated ways to debug active record:

https://user-images.githubusercontent.com/289882/167911312-4da245ae-30e7-4df6-b912-207e3d4a5ce3.mov

### Other Information

Async is a thing I have not looked at very much, I am not 100% sure how this should work in that context. I think the simplest first implementation is to pass the current state of the dynamic includes setting into the thread and just let the methods called in the thread do their thing. One other consideration here is how to handle debugging when threaded, how could we get all the loaded records back to the main thread and add them correctly as loaded state?

### Thanks for the help

I got feedback and help from a number of people prior to opening this pr. Thanks to @seejohnrun, @composerinteralia, @jhawthorn, @matthewd, @mtodd, @rushtonmd and @katestud for all the help, advice and guidance.

### CODE REVIEW

File: activerecord/lib/active_record/associations/association.rb [hunk 1]
```diff
@@ -228,15 +229,72 @@ def find_target
...
             end
+
+            ActiveRecord::LoadTree.add_loaded_records(records)
           end
+          records
+        end
+
+        def preloaded_records?
+          return false unless owner.respond_to?(reflection.name) # Sanity check
+          return false if owner.strict_loading_n_plus_one_only? # If strict loading is on do not preload
```
Comment (gwincr11) [inline activerecord/lib/active_record/associations/association.rb:259]: I am not positive this is the desired behavior.

File: activerecord/lib/active_record/associations/association.rb [hunk 2]
```diff
@@ -228,15 +229,72 @@ def find_target
...
+          return false if owner._load_tree.siblings.length <= 1 # no siblings so lets not go through the preload tooling,
+          # single records can sometime have optimizations not available
+          # to preloaders
+          true
+        end
+
+        def attempt_preload
+          available_records = [owner._load_tree.siblings]
+          available_records = Array.wrap(target) + available_records
+          available_records = available_records.flatten.uniq
```
Comment (gwincr11) [inline activerecord/lib/active_record/associations/association.rb:270]: I am not sure if these available records are actually being used, or rather when. It seems that there are a few guards around how to use this part of the api.

File: activerecord/lib/active_record/load_tree.rb [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    # Reset the state of tracked loaded and used records.
+    def self.clear_loaded_records
+      ActiveSupport::IsolatedExecutionState[:active_record_loaded_records] = []
+    end
+
+    # Create a Load Tree instance representing the loaded record and its place in the overall
+    # hierarchy of the object tree.
+    #
+    # parent: ActiveRecord instance that is the parent of the loaded record.
+    # siblings: ActiveRecord instances that were loaded at the same time as the loaded record.
```
Comment (gwincr11) [inline activerecord/lib/active_record/load_tree.rb:107]: These do not have to be ActiveRecord Instances.

File: activerecord/lib/active_record/load_tree.rb [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      @loaded_associations = []
+      @load_call_stack = caller if ActiveRecord.dynamic_includes_overfetch_debugging?
+      reset_siblings(siblings)
+    end
+
+    def reset_siblings(siblings)
+      # We only want to keep sibling of the same class type as the creator
+      # This ensures we will not try to preload any associations that may not
+      # exist on some of the sibling which were loaded because of STI.
+      @siblings = siblings.select { |s| s.class.name == @record_class_name }.compact
```
Comment (gwincr11) [inline activerecord/lib/active_record/load_tree.rb:127]: I was storing these as WeakRefs, but after a conversation with @jhawthorn I switched back to just directly referencing them.

File: activerecord/lib/active_record/dynamic_includes.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+      def self.disable_dynamic_includes
+        prior_value = self.dynamic_includes_enabled?
+        self.dynamic_includes_enabled = false
+        yield
+      ensure
+        self.dynamic_includes_enabled = prior_value
+      end
+
+      def self.dynamic_includes_overfetch_debugging=(value)
```
Comment (gwincr11) [inline activerecord/lib/active_record/dynamic_includes.rb:77]: I think there could be a better name for this mode if it sticks around.

File: activerecord/lib/active_record/load_tree.rb (hunk 4 above)
Comment (matthewd) [reply activerecord/lib/active_record/load_tree.rb:128]: I still think these should be weakrefs. Querying on behalf of an object that's no longer accessible from the live object graph seems pretty clearly wasteful?

File: activerecord/lib/active_record/associations/association.rb (hunk 1 above)
Comment (eileencodes) [reply activerecord/lib/active_record/associations/association.rb:259]: I don't think this is right, the purpose of strict_loading is to force devs to use preloading.

File: activerecord/lib/active_record/associations/association.rb [hunk 6]
```diff
@@ -228,15 +229,72 @@ def find_target
...
+
+          prs = ActiveRecord::Associations::Preloader.new(
```
Comment (eileencodes) [inline activerecord/lib/active_record/associations/association.rb:272]: Let's call this `preloaded_records` so that it's easier to know what's going on. Right now I read it as pull request 😆

File: activerecord/lib/active_record/dynamic_includes.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  # DynamicIncludes is a module that can be included to allow for dynamic loading of
```
Comment (eileencodes) [inline activerecord/lib/active_record/dynamic_includes.rb:4]: I'm kind of confused by this sentence. Is the intention for applications to be required to include this? I don't think that's necessary since this module is autoloaded in activerecord/lib/active_record.rb. If apps aren't supposed to include this or use it directly, it should be `no-doc`.

File: activerecord/lib/active_record/load_tree.rb [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    # checking if the current load tree is the the root, if it is then it returns
+    # the class name of the record for the root element, if not it returns the association
+    # name that was used to instantiate the current load tree's record. This will put together
+    # a full method chain.
+    #
+    # parent = Person.find(1)
+    # child = parent.children.first
+    # child._load_tree.full_load_call_path # => "Person.children"
+    #
+    def full_load_path
```
Comment (eileencodes) [inline activerecord/lib/active_record/load_tree.rb:146]: `full_load_path` makes me think of the file load path, maybe `full_load_tree`? I'm not sure but I keep getting hung up on this name.

File: activerecord/lib/active_record/load_tree.rb [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    # Add an association name to the list of loaded associations.
+    def add_loaded_association(association_name)
+      return if association_name.nil? || @loaded_associations.include?(association_name)
+      @loaded_associations << association_name
+    end
+
+    # Set the load trees parent to another ActiveRecord instance.
+    def set_parent(parent)
+      @parent = parent
+    end
```
Comment (eileencodes) [inline activerecord/lib/active_record/load_tree.rb:160]: Any methods in here, like `set_parent` that aren't called by external classes should be under the private namespace.

File: activerecord/lib/active_record/load_tree.rb [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+      raise ArgumentError, "all siblings must be have a load tree" unless siblings.all? { |r| r.respond_to?(:_load_tree) && r._load_tree.is_a?(LoadTree) }
+      set_parent(parent)
+
+      @record_class_name = creator.class.name
+      @association_name = association_name
+      parent._load_tree.add_loaded_association(association_name) unless parent.nil? || association_name.nil?
+      @loaded_associations = []
+      @load_call_stack = caller if ActiveRecord.dynamic_includes_overfetch_debugging?
+      reset_siblings(siblings)
+    end
```
Comment (eileencodes) [inline activerecord/lib/active_record/load_tree.rb:122]: From a design perspective I'm thinking this would be cleaner written like this:

```ruby
attr_reader :parent, :siblings, :record_class_name, :association_name, :loaded_association, :load_call_stack
def initialize(creator: nil, parent: nil, siblings: [], association_name: nil)
  @record_class_name = creator.class.name
  @association_name = association_name
  @loaded_associations = []
  @load_call_stack = caller if ActiveRecord.dynamic_includes_overfetch_debugging?
  @parent = parent
  @siblings = siblings
end

def set_records
  set_parent
  reset_siblings
end

etc
```

Then the caller would become `ActiveRecord::LoadTree.new(creator: self, siblings: siblings, parent: parent, association_name: association_name).set_records`. The changes here ensure initialize is responsible only for initialization, then the methods below make each of their roles clear. By using the attr_reader we now make the parent/siblings etc accessible so we don't need to either pass the parent/sibling OR access them with @. The goal here is a cleaner interface.

File: activerecord/test/cases/load_tree_test.rb [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+require "models/company"
+require "models/computer"
+require "models/mentor"
+require "models/project"
+require "models/ship"
+require "models/ship_part"
+require "models/strict_zine"
+require "models/interest"
+require "models/treasure"
+require "models/pirate"
```
Comment (eileencodes) [inline activerecord/test/cases/load_tree_test.rb:15]: I think some of these aren't needed, can you remove the ones that you aren't using in here? At a quick glance, pirate, treasure, and interest aren't used.

**Review by eileencodes — CHANGES_REQUESTED** [review]
Hi @gwincr11! This is an interesting idea, thanks for opening a PR.

I've left some comments below, some about style others about docs. My main concern is the idea that applications can nest multiple calls in and out of preloading. The resulting code is hard to read and reason about, I wonder if that API could be improved (or potentially not allowed). Public APIs, what apps consume, are important to us so we want to make sure that it's intuitive, clean, and the intention is clear. 

Some general notes about the PR:

1) It needs a changelog and docs for the public interface. Right now the private interfaces and behaviors are documented but we don't want apps consuming those.
2) PRs should have only one commit, can you squash these? the 71 commits are taking up a lot of timeline right now, and you'll need to squash before merge anyway.
3) Is there anyway to reduce the surface area of this PR? I'm thinking for starters we remove the overfetching debugging. I'm asking this so that it's easier to review the code. We can open a future PR to add overfetching debugging if deemed necessary.

File: activerecord/lib/active_record/associations/association.rb (hunk 6 above)
Comment (gwincr11) [reply activerecord/lib/active_record/associations/association.rb]: I called it preloaders, since it is not really records but an array of preloaders that ran. Do you like that?

File: activerecord/lib/active_record/load_tree.rb (hunk 8 above)
Comment (gwincr11) [reply activerecord/lib/active_record/load_tree.rb:71]: I had not idea what to call this either, naming is hard... Part of me thinks it should be something having to do with a tree or a graph, because it is sort of like its place in a family tree, but I have no idea what that would be. Otherwise maybe something about methods called?

File: activerecord/lib/active_record.rb [hunk 12]
```diff
@@ -44,12 +44,14 @@ module ActiveRecord
   autoload :DelegatedType
   autoload :DestroyAssociationAsyncJob
   autoload :DynamicMatchers
+  autoload :DynamicIncludes
```
[the remark this answers is not in the document]
Comment (gwincr11) [reply activerecord/lib/active_record.rb:47]: I originally called it `enable_auto_includes` however @matthewd said that `dynamic` was a better name. I am cool with either choice.

[the remark this answers is not in the document]
Comment (eileencodes) [reply activerecord/lib/active_record.rb:47]: Ha, well then I guess Matthew and I will have to argue about it in our next pairing session 😆

**Comment by gwincr11** [comment]
@eileencodes I have removed overfetch debugging, I think it is pretty helpful do you think I should open a separate pr?

File: activerecord/lib/active_record/load_tree.rb (hunk 10 above)
Comment (gwincr11) [reply activerecord/lib/active_record/load_tree.rb:27]: Should the error checking still occur in the initialize statement?

Comment (eileencodes) [reply activerecord/lib/active_record/load_tree.rb:27]: I think the error checking is overkill to be honest. The interface should inform the type of data sent, and since apps shouldn't be initializing their own LoadTree, that initializer isn't the right place to handle it anyway. Basically the LoadTree initializer should always be sent the right data from the caller (in this case wherever it's initialized in AR).

**Comment by gwincr11** [comment]
I am going to break this pr up into 3 separate prs. 

1. Load Tree
2. Dynamic Includes
3. Debugging Tools
